### DISCUSSION TITLE
Custom Functions

### DISCUSSION THREAD

**Discussion by lvca** [ArcadeData/arcadedb#281] [Ideas]
OrientDB supports custom functions written in any language supported by the JVM (mostly Javascript using Nashorn), so other DBMSs allow the user to create their own functions you can call from the query language or HTTP.

In ArcadeDB we could support this feature, by using the underlying advanced GraalVM infrastructure that allows to compile and execute multiple scripting languages on top of the JVM. We did this job already with Arcade Trader (https://arcadetrader.com) with a minimal IDE to edit the functions, syntax highlight, inline documentation, etc. We could totally port that piece of IDE into Studio.

Requirements in order of implementation and status update:

- [x] Custom Functions can be written in multiple languages, Javascript and Java must be supported in the first release
- [x] Java and Polyglot query engines must protect from the execution of any Java classes
- [x] The execution of a function should be limited with a timeout. This avoids also blocking by writing a loop such as `while(true);`
- [x] New API to register functions. Examples:
 - Javascript `database.getSchema().registerFunction("js", "sum", "function sum(a,b){ return a + b; }");` and 
 - Java `database.getSchema().registerFunction("java", "sum", "org.acme.Functions::sum");`
- [x] Create SQL statements to register and unregister functions
- [ ] Allow executing a SQL script to make the registration easy at startup with a new setting `arcadedb.startupScript`. Example: `-Darcadedb.startupScript=/temp/myscript.sql`
- [ ] Custom Functions should be automatically published on HTTP if requested by the user (in the IDE a select field with ["no published", "HTTP get", "HTTP post"])
- [ ] Create a tutorial about writing a recommendation system by using a custom function

They are **not Stored Procedures**, because they are not stored. If you want to use them, register in the database at startup or before using them. This avoids all the issues with where to store them (files, internet types/buckets?), security about who can write, read and execute them, replication, an editor from the studio, and more.

**Comment by lvca** [comment]
Created issue #288 and the work in progress will be on the branch https://github.com/ArcadeData/arcadedb/compare/custom-functions-poc.

**Reply by Walker555** [reply]
Hi Luca. I try to migrate OrientDB 3.0.x queries to ArcadeDB. The queries use a lot of http://orientdb.com/docs/3.0.x/sql/SQL-Functions.html
The `SELECT if(eval( .... etc.` isn't supported by ArcadeDB. Is there a way to migrate the functions by writing custom Java functions? and put the custom JAR on classloader path?

**Reply by tolgaulas** [reply]
That's interesting, because we use eval too here and there in orientdb but recently it was decided to remove eval in ArcadeDB, but we're not alone as it seems, may be @lvca review that decision :).  (See #563 )

**Reply by lvca** [reply]
Hi guys, I didn't know eval was so used! @Walker555 could you please write down a query that used the `eval()` in OrientDB? Same for @tolgaulas if you can. Thanks.

**Reply by tolgaulas** [reply]
Here is an innocent one from odb : `select eval("'100000.123'.toFloat()").asDecimal()*1000` from a lazy dev. I admit its not the correct way to put it as it can be directly casted. We don't have chance to review all code base as they are somehow will be put in production in a not so much top-notch codebase environment.   I actually second removal of eval, but knowing a good ugly tool is there despite I prevent to use it, is a good assuring feeling.

**Reply by lvca** [reply]
You can execute the same query as `select "100000.123".asDecimal()*1000`.

**Reply by Walker555** [reply]
@lvca The original OrientDB is pretty long, I put only an anonymized abstract in here:
`SELECT if(eval('edge_a.credit_a >= 100'), edge_a.credit_a, edge_a.credit_b) as edge_a_credit ... order by edge_a_credit desc`

**Comment by lvca** [comment]
Closing this discussion as Part 1. Part 2 is on https://github.com/ArcadeData/arcadedb/discussions/1155.
